osutil/strace/timing.go

```diff
@@ -0,0 +1,225 @@
+var execveatRE = regexp.MustCompile(`([0-9]+)\ +([0-9.]+) execveat\(.*\["([^"]+)"`)
+
+// lines look like (both SIGTERM and SIGCHLD need to be handled):
+// PID   TIME                  SIGNAL
+// 17559 1542815330.242750 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=17643, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
+var sigChldTermRE = regexp.MustCompile(`[0-9]+\ +([0-9.]+).*SIG(CHLD|TERM)\ {.*si_pid=([0-9]+),`)
+
+// all lines start with this:
+// PID   TIME
+// 21616 1542882400.198907 ....
+var timeRE = regexp.MustCompile(`[0-9]+\ +([0-9.]+).*`)
+
+func handleExecMatch(trace *ExecveTiming, pt *pidTracker, match []string) error {
+	if len(match) == 0 {
+		return nil
+	}
+	// the pid of the process that does the execve{,at}()
+	pid := match[1]
+	execStart, err := strconv.ParseFloat(match[2], 64)
+	if err != nil {
+		return err
+	}
+	exe := match[3]
+	// deal with subsequent execve()
+	if start, exe := pt.Get(pid); exe != "" {
+		trace.addExeRuntime(exe, execStart-start)
+	}
+	pt.Add(pid, execStart, exe)
+	return nil
+}
+
+func handleSignalMatch(trace *ExecveTiming, pt *pidTracker, match []string) error {
+	if len(match) == 0 {
+		return nil
+	}
+	sigTime, err := strconv.ParseFloat(match[1], 64)
+	if err != nil {
+		return err
+	}
+	sigPid := match[3]
+	if start, exe := pt.Get(sigPid); exe != "" {
+		trace.addExeRuntime(exe, sigTime-start)
+		pt.Del(sigPid)
+	}
+	return nil
+}
+
+func TraceExecveTimings(straceLog string, nSlowest int) (*ExecveTiming, error) {
+	slog, err := os.Open(straceLog)
+	if err != nil {
+		return nil, err
+	}
+	defer slog.Close()
+
+	// pidTracker maps the "pid" string to the executable
+	pidTracker := newPidTracker()
+
+	var line string
+	var start, end, tmp float64
+	trace := NewExecveTiming(nSlowest)
+	r := bufio.NewScanner(slog)
+	for r.Scan() {
+		line = r.Text()
+		if start == 0.0 {
+			fmt.Sscanf(line, "%f %f ", &tmp, &start)
+		}
+		// handleExecMatch looks for execve{,at}() calls and
+		// uses the pidTracker to keep track of execution of
+		// things. Because of fork() we may see many pids and
+		// within each pid we can see multiple execve{,at}()
+		// calls.
+		// An example of pids/exec transitions:
+		// $ snap run --trace-exec test-snapd-sh -c "/bin/true"
+		//    pid 20817 execve("snap-confine")
+		//    pid 20817 execve("snap-exec")
+		//    pid 20817 execve("/snap/test-snapd-sh/x2/bin/sh")
+		//    pid 20817 execve("/bin/sh")
+		//    pid 2023  execve("/bin/true")
+		match := execveRE.FindStringSubmatch(line)
+		if err := handleExecMatch(trace, pidTracker, match); err != nil {
+			return nil, err
+		}
+		match = execveatRE.FindStringSubmatch(line)
+		if err := handleExecMatch(trace, pidTracker, match); err != nil {
+			return nil, err
+		}
+		match = sigChldTermRE.FindStringSubmatch(line)
+		if err := handleSignalMatch(trace, pidTracker, match); err != nil {
```

Comment:
could we add a comment about the role of this as well as we did for handleExecMatch